_posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd

```diff
@@ -0,0 +1,681 @@
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/traceplot_sigma.png")
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/traceplot_a_1.png")
+```
+
+
+How about the posterior distributions of the parameters, first and foremost, the varying intercepts `a_1` - `a_48`?
+
+### Posterior distributions
+
+```{r}
+plot_posterior <- function(samples) {
+  prep_tibble(samples) %>% 
+    ggplot(aes(x = value, color = chain)) +
+    geom_density() +
+    theme_classic() +
+    theme(legend.position = "none",
+          axis.title = element_blank(),
+          axis.text = element_blank(),
+          axis.ticks = element_blank())
+    
+}
+
+plot_posteriors <- function(sample_array, num_params) {
+  plots <- vector(mode = "list", length = num_params)
+  for (i in 1:num_params) {
+    p <- plot_posterior(sample_array[ , , i] %>% as.matrix())
+    plots[[i]] <- p
+  }
+  do.call(grid.arrange, plots)
+}
+
+plot_posteriors(mcmc_trace[[3]], dim(mcmc_trace[[3]])[3])
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/posteriors.png")
+```
+
+Now let's see the corresponding posterior means and highest posterior density intervals.
+(The below code includes the hyperpriors in `summary` as we'll want to display a complete _precis_-like output soon.)
+
+### Posterior means and HPDIs
+
+```{r}
+all_samples <- all_samples %>%
+  as_tibble(.name_repair = ~ c("a_bar", "sigma", paste0("a_", 1:48))) 
+
+means <- all_samples %>% 
+  summarise_all(list (~ mean)) %>% 
+  gather(key = "key", value = "mean")
+
+sds <- all_samples %>% 
+  summarise_all(list (~ sd)) %>% 
+  gather(key = "key", value = "sd")
+
+hpdis <-
+  all_samples %>%
+  summarise_all(list(~ list(hdi(.) %>% t() %>% as_tibble()))) %>% 
+  unnest() 
+
+hpdis_lower <- hpdis %>% select(-contains("upper")) %>%
+  rename(lower0 = lower) %>%
+  gather(key = "key", value = "lower") %>% 
+  arrange(as.integer(str_sub(key, 6))) %>%
+  mutate(key = c("a_bar", "sigma", paste0("a_", 1:48)))
+
+hpdis_upper <- hpdis %>% select(-contains("lower")) %>%
+  rename(upper0 = upper) %>%
+  gather(key = "key", value = "upper") %>% 
+  arrange(as.integer(str_sub(key, 6))) %>%
+  mutate(key = c("a_bar", "sigma", paste0("a_", 1:48)))
+
+summary <- means %>% 
+  inner_join(sds, by = "key") %>% 
+  inner_join(hpdis_lower, by = "key") %>%
+  inner_join(hpdis_upper, by = "key")
+
+
+summary %>% 
+  filter(!key %in% c("a_bar", "sigma")) %>%
+  mutate(key_fct = factor(key, levels = unique(key))) %>%
+  ggplot(aes(x = key_fct, y = mean, ymin = lower, ymax = upper)) +
+   geom_pointrange() + 
+   coord_flip() +  
+   xlab("") + ylab("post. mean and HPDI") +
+   theme_minimal() 
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/forest_plot.png")
+```
+
+Now for an equivalent to _precis_. We already computed means, standard deviations and the HPDI interval.
```

Comment:
It now occurs to me you could be referring to something technical I don't know about called _precis_, but if just the word, I use an accent. Looks like it's dealer's choice in U.S. English, though.

Reply:
yeah exactly, it's been "coined" by McElreath and he writes it without the accent ... (((normally I'm snobbish enough to leave them in when I write English && use some French word, but here really I shouldn't ;-)))